### ISSUE TITLE
WC_Coupon::get_coupon_id_from_code() SQL optimization

### ISSUE THREAD

**Issue by pmgarman** [woothemes/woocommerce#9902]
![](http://c.pmgr.mn/HEx2+)

This is a brief snippet of a trace of an `update_order_review`  wc-ajax call. The order ultimately had just `1` coupon attached to it, so I'm not sure why it tried pulling 14 coupons, but it did. I'll dig into that later. However each time a coupon is pulled from the DB it's about 800-1100 seconds on average (for us) through most of the traces I look at. This is almost entirely a single SQL query.

`get_coupon_id_from_code()` will query the posts table for a `shop_coupon` post type that is published with the `post_title` of the coupon code being searched for. The `post_title` column is not index. This creates slow queries. The `post_name` column *is indexed* BUT the `post_title` and `post_name` are not guaranteed to be the same. Specifically for us `0.114%` of coupons they are not the same. However most of our coupons getting generated are made programatically making the chance for error much smaller. Any chance of error means just using `post_name` for the query though is not a great idea.

I think we can optimize this by moving this query by default using the `post_name` column getting us queries that run in maybe `5.5%` the time. However, what can we do on the code side to ensure this always matches the `post_title` column? I'm likely going to filter the query myself to just start running off that and use SQL to fix our 411 mismatched titles/names but this also means I need to monitor this to ensure it always stays in sync vs being able to trust they are in sync.

#### Some MySQL Queries to give some context and data

Total coupons in the site
```
mysql> select count(*) from wp_posts where post_type = 'shop_coupon' and post_status = 'publish';
+----------+
| count(*) |
+----------+
|   361466 |
+----------+
1 row in set (0.32 sec)
```

Total published posts within the site
```
mysql> select count(*) from wp_posts where post_status = 'publish';
+----------+
| count(*) |
+----------+
|   446173 |
+----------+
1 row in set (0.65 sec)
```

Total coupons in site with mismatching title or name
```
mysql> select count(*) from wp_posts where post_type = 'shop_coupon' and post_status = 'publish' and post_title != post_name;
+----------+
| count(*) |
+----------+
|      411 |
+----------+
1 row in set (1.12 sec)
```

Query currently being used searching for a specific coupon
```
mysql> select ID from wp_posts where post_title = '?' and post_type = 'shop_coupon' and post_status = 'publish';
+---------+
| ID      |
+---------+
| ? |
+---------+
1 row in set (1.09 sec)
```

Optimized query using the post_name currently being used searching for a specific coupon
```
mysql> select ID from wp_posts where post_name = '?' and post_type = 'shop_coupon' and post_status = 'publish';
+---------+
| ID      |
+---------+
| ? |
+---------+
1 row in set (0.06 sec)
```

**Comment by pmgarman** [comment]
To sync up the coupons post titles and names I've created this CLI command. This will if the cleaned title and post title (so the coupon code and sanitized coupon code) match, it will update the post name with the sanitized post title, making lookup by post name possible for that coupon code. If the sanitized post title and post title do *not* match, it will throw a warning and do nothing, unless you pass the `--force` argument and then it breaks the coupon by updating the post title and post name to the sanitized version.

```
		public function sync_coupon_slugs( $args, $assoc_args ) {
			global $wpdb;

			$coupons = $wpdb->get_results( "select ID, post_title, post_name from {$wpdb->posts} where post_type = 'shop_coupon' and post_status = 'publish' and post_title != post_name" );

			if( empty( $coupons ) ) {
				return;
			}

			foreach( $coupons as $coupon ) {
				$clean_title = sanitize_title( $coupon->post_title );

				if( $clean_title == $coupon->post_title ) {

					wp_update_post( array(
						'ID'        => $coupon->ID,
						'post_name' => $clean_title
					) );

				} else {
					if( isset( $assoc_args['force'] ) && true == $assoc_args['force'] ) {
						wp_update_post( array(
							'ID'         => $coupon->ID,
							'post_title' => $clean_title,
							'post_name'  => $clean_title
						) );
						add_post_meta( $coupon->ID, 'cp_force_sync', time(), true );
					} else {
						WP_CLI::line( sprintf( __( 'Could not update coupon ID %1$d - %2$s', 'colourpop' ), $coupon->ID, $coupon->post_title ) );
					}
				}
			}
		}
```

After fixing all the coupon codes that were not going to work with using the post_name I put the below filter in place. Typicaly many of our traces in New Relic would be ajax calls updating orders or apply coupons because of getting coupons and populating them. Since this filter being put in place, I've actually not been able to find any traces that were slow due to coupons.

```
function pmgarman_wc_coupon_code_query( $sql ) {
	global $wpdb;

	$sql = "SELECT ID FROM {$wpdb->posts} WHERE post_name = %s AND post_type = 'shop_coupon' AND post_status = 'publish'";

	return $sql;
}
add_filter( 'woocommerce_coupon_code_query', 'pmgarman_wc_coupon_code_query' );
```

**Comment by mikejolley** [comment]
If this is going to break coupons in a way the average user cannot correct it (it's not easy to set post_name), I don't think it should be changed.

2.5 adds wp_cache usage to get_coupon_id_from_code - didn't that help your case at all?

**Comment by pmgarman** [comment]
I have to imagine that through code during the creation and updating of coupon codes the post title and post name could be synced up. Currently there is no way to edit the post name (slug) when editing the coupon so the average user (and I'm sure quite a few developers) don't even realize "hey yeah its there behind the scenes)"

While the caching does help on subsequent lookups, that first lookup is still going to be hard. If there isn't a good way (which I see as totally valid) to keep the post title and post name in sync reliable and without breaking things I agree it should be left as is and I can just turn this into a quick plugin if anyone else wanted to use it. Something pretty minor with such a big win in an area of WooCommerce that has been plaquing this large site in performance figured it was at least worth throwing up an issue to see if we could improve on it.

**Comment by mikejolley** [comment]
@pmgarman have you traced back why you had multiple queries at once at all? That would obviously have more impact that 1 lookup (with cache).

**Comment by pmgarman** [comment]
I definitely agree, I think I know what it was but I have no way to validate it. We send out coupons for signing up to an email newsletter and its attached to that email it was sent to. Some _smart_ people decided "lets make a bunch of emails and get a bunch of coupons and use them all at the same time" ... they all find out on checkout when a bunch of "This coupon doens't belong to you!" notices during checkout.

To get around this we've set all these coupons to now be individual use, so they won't be able to stack so many coupons on a single order, and we can avoid the 14 coupon lookups.

Since I've not been able to reproduce WC doing this with only one coupon, this is my best guess as to why there were so many coupon lookups, after seeing the same thing with about 6 coupons on one order.

While WooCommerce does allow for an infinite number of coupons to be added to the order, thats not really a _bug_ in WooCommerce needing fixed, it's the stores responsibility on that part. So the 14 queries was on us. I'll go ahead and close this, though I may consider still putting the code I optimized all the stuff with into a plugin at some point.
